### Цели и задачи проекта

- **Общая подготовка данных к анализу**
  - причёсывание датасета;
  - конвертация типов данных в столбцах из `object` к оптимальным типам;
  - исправление пропусков, чтобы минимизировать влияние на результат и сохранить строки с остальными полезными полями;
  - удаление дубликатов строк;
  - объединение неявных дубликатов.


- **Специальная подготовка данных к анализу (превращение в DataMart)**
  - фильтрация по заданным критериям, нужным в будущем анализе;
  - кластеризация данных по рейтингу пользователей и критиков.

# Анализ продаж игр

- Автор: Федоров Алексей Анатольевич
- Дата: 31.03.2026

 ### Описание данных

- **Name** — название игры.  
- **Platform** — название платформы.  
- **Year of Release** — год выпуска игры.  
- **Genre** — жанр игры.  
- **NA sales** — продажи в Северной Америке (в миллионах проданных копий).  
- **EU sales** — продажи в Европе (в миллионах проданных копий).  
- **JP sales** — продажи в Японии (в миллионах проданных копий).  
- **Other sales** — продажи в других странах (в миллионах проданных копий).  
- **Critic Score** — оценка критиков (от 0 до 100).  
- **User Score** — оценка пользователей (от 0 до 10).  
- **Rating** — рейтинг организации ESRB (англ. *Entertainment Software Rating Board*), которая присваивает играм возрастные категории.

Справочник Rating:
| Рейтинг | Описание | Возрастная категория | Период действия |
| :--- | :--- | :--- | :--- |
| **RP** | **Rating Pending** — ожидается окончательный рейтинг. | Любой | 1994 — н.в. |
| **EC** | **Early Childhood** — для дошкольников. Без неприемлемого контента. | 3+ | 1994 — 2018 |
| **E** | **Everyone** — для всех (до 1998 г. назывался **K-A**). | 6+ | 1994 — н.в. |
| **E10+** | **Everyone 10+** — умеренное насилие и грубая речь. | 10+ | 2005 — н.в. |
| **T** | **Teen** — для подростков. Допустимы кровь и сексуальный подтекст. | 13+ | 1994 — н.в. |
| **M** | **Mature** — для лиц от 17 лет. Реалистичное насилие и нагота. | 17+ | 1994 — н.в. |
| **AO** | **Adults Only** — строго для взрослых. Высокий уровень воздействия. | 18+ | 1994 — н.в. |

#### Подробные описания:

*   **RP (Rating Pending)** — используется в рекламных материалах для игр, которым ещё не присвоен окончательный рейтинг ESRB. Принят в 1994 году и действует до сих пор.
*   **EC (Early Childhood)** — игры, ориентированные на дошкольную аудиторию. Не содержат материалов, которые родители могли бы считать неприемлемыми. Принят в 1994 году и действовал до 2018 года.
*   **E (Everyone)** — игры для всех возрастов. Могут содержать незначительное количество насилия и сленга. До 1998 года рейтинг был известен как Kids to Adults (K-A). Принят в 1994 году и действует до сих пор.
*   **E10+ (Everyone 10 and older)** — игры для людей от 10 лет и старше. Могут содержать более высокий уровень насилия, грубую разговорную речь. Принят в 2005 году и действует до сих пор.
*   **T (Teen)** — игры для людей от 13 лет и старше. Могут содержать умеренное количество насилия, включая небольшие количества крови, грубую разговорную речь и юмор, сексуальное содержание. Принят в 1994 году и действует до сих пор.
*   **M (Mature)** — игры для лиц от 17 лет и старше. Могут содержать более интенсивные или реалистичные изображения насилия, включая кровь, ужасы, увечья и изображения смерти, непристойные сексуальные сцены, частичную наготу и грубую ненормативную лексику. Принят в 1994 году и действует до сих пор.
*   **AO (Adults Only)** — игры для людей от 18 лет и старше. Содержат материалы с более высоким уровнем воздействия, чем рейтинг M. Принят в 1994 году и действует до сих пор.

 ### Содержимое проекта

<div style="background-color: #f8f9fa; padding: 10px; border-left: 3px solid #007acc;">
    <h3>Оглавление</h3>
    <ul>
        <li><a href="#1-bullet">Загрузка данных и знакомство с ними</a>
            <ul>                                 
                <li><a href="#1-2-bullet">Проверка корректности и составление справочников Platform, Genre, Rating, User Score, Critic Score</a></li>
                <li><a href="#1-3-bullet">Выводы по датасету</a></li>                
                <li><a href="#1-4-bullet">Чистовая загрузка данных</a></li>                
            </ul>
        </li>        
        <li><a href="#2-bullet">Проверка ошибок в данных и их предобработка</a>
            <ul>
                <li><a href="#2-1-bullet">Названия, или метки, столбцов датафрейма</a></li>
                <li><a href="#2-2-bullet">Типы данных</a></li>
                <li><a href="#2-3-bullet">Наличие пропусков в данных</a></li>
                <li><a href="#2-4-bullet">Явные и неявные дубликаты в данных</a></li>                
            </ul>
        </li>
        <li><a href="#3-bullet">Фильтрация данных</a></li>
        <li><a href="#4-bullet">Категоризация данных</a></li>        
    </ul>
</div>

---

<a class='anchor' id='1-1-bullet'></a>
 ### Черновая загрузка датасета

<a class='anchor' id='1-bullet'></a>
## Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [1]:
import pandas as pd
import sys
import os
print(sys.executable) # проверяем, что подцепилось ядро Jupyter с нужным venv
print(os.getcwd()) # смотрим текущую папку, чтобы понимать, как указывать путь к CSV

D:\Learn\YandexPracticum\DataScience1\sprint6_pandas\.venv\Scripts\python.exe
D:\Learn\YandexPracticum\DataScience1\sprint6_pandas


In [2]:
def show(num, name, func):
    print('\n', f'{num}. {name}', '-' * 20, '\n')
    func()

def get_missings(df):
    missing = df.isna().sum().loc[lambda v: v > 0].sort_values(ascending=False)
    missing_fraction = round(missing / df.shape[0] * 100, 2)
    return pd.DataFrame({'N/A': missing, 'Fraction %': missing_fraction})

def get_reference_info(df, name, numeric=False):
    values = df[name].drop_duplicates().sort_values()
    lst = values.tolist()
    print(f'{name}: ', lst[:150], len(lst), end='\n\n')
    print(df[name].value_counts(dropna=False), end='\n\n')
    if numeric:
        values_num = pd.to_numeric(values, errors='coerce')
        print(f'min = {values_num.min()}, max = {values_num.max()}\n\n')


In [3]:
df = pd.read_csv('./datasets/new_games.csv') # черновая загрузка датасета

# ознакомление со структурой

print(f'Всего: {df.shape[0]} строк, {df.shape[1]} колонок')
print(f'Колонки: {df.columns.tolist()}')

show(1, 'info', lambda: df.info())
show(2, 'missing', lambda: display(get_missings(df)))
show(3, 'describe', lambda: print(df.describe()))
show(4, 'display', lambda: display(df))

Всего: 16956 строк, 11 колонок
Колонки: ['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales', 'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating']

 1. info -------------------- 

<class 'pandas.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  str    
 1   Platform         16956 non-null  str    
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  str    
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  str    
 6   JP sales         16956 non-null  str    
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  str    
 10  Rating           10085 non-null  str    
dtypes: float64(4), str(7)
memory usage: 1.4 MB

 2. missing -------------------- 



,N/A,Fraction %
Critic Score,8714,51.39
Rating,6871,40.52
User Score,6804,40.13
Year of Release,275,1.62
Name,2,0.01
Genre,2,0.01



 3. describe -------------------- 

       Year of Release      NA sales   Other sales  Critic Score
count     16681.000000  16956.000000  16956.000000   8242.000000
mean       2006.485522      0.262023      0.047087     68.926717
std           5.873102      0.808654      0.185577     13.944565
min        1980.000000      0.000000      0.000000     13.000000
25%        2003.000000      0.000000      0.000000     60.000000
50%        2007.000000      0.080000      0.010000     71.000000
75%        2010.000000      0.240000      0.030000     79.000000
max        2016.000000     41.360000     10.570000     98.000000

 4. display -------------------- 



,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...
16951,Samurai Warriors: Sanada Maru,PS3,2016.0,Action,0.00,0.0,0.01,0.00,NaN,NaN,NaN
16952,LMA Manager 2007,X360,2006.0,Sports,0.00,0.01,0.0,0.00,NaN,NaN,NaN
16953,Haitaka no Psychedelica,PSV,2016.0,Adventure,0.00,0.0,0.01,0.00,NaN,NaN,NaN
16954,Spirits & Spells,GBA,2003.0,Platform,0.01,0.0,0.0,0.00,NaN,NaN,NaN


<a class='anchor' id='1-2-bullet'></a>
 ### Проверка корректности и составление справочников Platform, Genre, Rating, User Score, Critic Score

In [4]:
for col_name, num_flag in (('Platform', False), ('Genre', False), ('Rating', False), ('User Score', True), ('Critic Score', True), ('Year of Release', True)):
    get_reference_info(df, col_name, num_flag)

Platform:  ['2600', '3DO', '3DS', 'DC', 'DS', 'GB', 'GBA', 'GC', 'GEN', 'GG', 'N64', 'NES', 'NG', 'PC', 'PCFX', 'PS', 'PS2', 'PS3', 'PS4', 'PSP', 'PSV', 'SAT', 'SCD', 'SNES', 'TG16', 'WS', 'Wii', 'WiiU', 'X360', 'XB', 'XOne'] 31

Platform
PS2     2189
DS      2177
PS3     1355
Wii     1340
X360    1281
PSP     1229
PS      1215
PC       990
XB       839
GBA      837
GC       563
3DS      530
PSV      435
PS4      395
N64      323
XOne     251
SNES     241
SAT      174
WiiU     147
2600     135
NES      100
GB        98
DC        52
GEN       29
NG        12
SCD        6
WS         6
3DO        3
TG16       2
GG         1
PCFX       1
Name: count, dtype: int64

Genre:  ['ACTION', 'ADVENTURE', 'Action', 'Adventure', 'FIGHTING', 'Fighting', 'MISC', 'Misc', 'PLATFORM', 'PUZZLE', 'Platform', 'Puzzle', 'RACING', 'ROLE-PLAYING', 'Racing', 'Role-Playing', 'SHOOTER', 'SIMULATION', 'SPORTS', 'STRATEGY', 'Shooter', 'Simulation', 'Sports', 'Strategy', nan] 25

Genre
Action          3405
Sports    

- Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.
- Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

<a class='anchor' id='1-3-bullet'></a>
 ### Выводы по датасету

<div style="width: 100%;">
  <table style="width: 100%; border-collapse: collapse;">
    <tr>
      <th style="width: 70px; min-width: 70px; text-align: center; border: 1px solid #ccc; padding: 6px;">№</th>
      <th style="width: auto; text-align: left; border: 1px solid #ccc; padding: 6px;">Аспект данных, проблемы</th>
      <th style="width: auto; text-align: left; border: 1px solid #ccc; padding: 6px;">Устранение</th>
    </tr>
    <tr>
      <td style="border: 1px solid #ccc; padding: 6px;">1</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Все колонки на месте: 11 штук, и почти 17 000 строк. Названия не в Snake Case.</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Надо переименовать часть колонок.</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ccc; padding: 6px;">2</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Строковые колонки имеют неоптимальные типы str.</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Надо привести к string и дополнительно обработать косвенные пропуски (none, unknown...).</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ccc; padding: 6px;">3</td>
      <td style="border: 1px solid #ccc; padding: 6px;">User Score должен быть Float64</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Задать конверсию типа и обработку косвенных пропусков</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ccc; padding: 6px;">4</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Critic Score должен быть Int64.</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Задать конверсию типа и обработку косвенных пропусков</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ccc; padding: 6px;">5</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Все поля Sales должны быть Float64.</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Задать конверсию типа и обработку косвенных пропусков</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ccc; padding: 6px;">6</td>
      <td style="border: 1px solid #ccc; padding: 6px;">При анализе значений обнаружилось, что есть косвенные пропуски: 'tbd', 'none', ' ', 'unknown'</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Их нужно указать DataFrame в na_values. Ещё могут быть пробелы у чисел в начале строки. Надо указать skipinitialspace.</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ccc; padding: 6px;">7</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Есть пропуски. Rating и все Score примерно 50% пропусков. year_of_release 1.62% и ещё eu_sales, jp_sales, genre, name имеют штучные пропуски менее 0.05%</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Решать по задаче. Где &lt; 5% можно удалить строки. Штучные без названия бесполезны и можно удалить. Это мусор. По рейтингам: maen, sum, count умеют пропускать такие строки. На данном этапе удалим только с пустым name и genre. Это безполезные строки.</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ccc; padding: 6px;">8</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Справочник Platforms качественный. В данных платформа 100% указана.</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Стоит поинтересоваться редкими: SCD, WS, 3DO, TG16, GG, PCFX. Возможно опечатки.</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ccc; padding: 6px;">9</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Справочник Genre некачественный: полно дубликатов. В данных 2 пропуска.</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Слить дубликаты в один через lowercase. Пропуски удалить со строкаами.</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ccc; padding: 6px;">10</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Справочник Rating качественный, но есть 1 лишний элемент, встречается 3 раза. 40.5% пропусков в данных.</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Уточнить про рейтинг К-А</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ccc; padding: 6px;">11</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Значения Scores качественные, без эксцессов, около 50% пропуски в данных.</td>
      <td style="border: 1px solid #ccc; padding: 6px;">Индивидуально по задаче.</td>
    </tr>
  </table>
</div>

<a class='anchor' id='1-4-bullet'></a>
 ### Чистовая загрузка данных

In [5]:
data_types = {'Year of Release': 'Int64', 'Critic Score': 'Int64', 
              'NA sales': 'Float64', 'EU sales': 'Float64', 'JP sales': 'Float64', 'Other sales': 'Float64',
              'User Score': 'Float64',
              'Name': 'string',
              'Platform': 'string',
              'Genre': 'string',
              'Rating': 'string',
             }
df = pd.read_csv('./datasets/new_games.csv', dtype=data_types, na_values=['tbd', 'none', ' ', 'unknown',], skipinitialspace=True).rename(columns=lambda col_name: col_name.replace(' ', '_').lower())

print(f'Колонки: {df.columns.tolist()}')

show(1, 'info', lambda: df.info())
show(2, 'missing', lambda: display(get_missings(df)))
show(3, 'describe', lambda: print(df.describe()))
show(4, 'display', lambda: display(df))

Колонки: ['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales', 'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating']

 1. info -------------------- 

<class 'pandas.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  string 
 1   platform         16956 non-null  string 
 2   year_of_release  16681 non-null  Int64  
 3   genre            16954 non-null  string 
 4   na_sales         16956 non-null  Float64
 5   eu_sales         16950 non-null  Float64
 6   jp_sales         16952 non-null  Float64
 7   other_sales      16956 non-null  Float64
 8   critic_score     8242 non-null   Int64  
 9   user_score       7688 non-null   Float64
 10  rating           10085 non-null  string 
dtypes: Float64(5), Int64(2), string(4)
memory usage: 1.5 MB

 2. missing -------------------- 



,N/A,Fraction %
user_score,9268,54.66
critic_score,8714,51.39
rating,6871,40.52
year_of_release,275,1.62
eu_sales,6,0.04
jp_sales,4,0.02
genre,2,0.01
name,2,0.01



 3. describe -------------------- 

       year_of_release  na_sales  eu_sales  jp_sales  other_sales  \
count          16681.0   16956.0   16950.0   16952.0      16956.0   
mean       2006.485522  0.262023  0.144114   0.07717     0.047087   
std           5.873102  0.808654  0.500305  0.307148     0.185577   
min             1980.0       0.0       0.0       0.0          0.0   
25%             2003.0       0.0       0.0       0.0          0.0   
50%             2007.0      0.08      0.02       0.0         0.01   
75%             2010.0      0.24      0.11      0.04         0.03   
max             2016.0     41.36     28.96     10.22        10.57   

       critic_score  user_score  
count        8242.0      7688.0  
mean      68.926717    7.122607  
std       13.944565    1.499804  
min            13.0         0.0  
25%            60.0         6.4  
50%            71.0         7.5  
75%            79.0         8.2  
max            98.0         9.7  

 4. display -------------------- 


,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,Wii Sports,Wii,2006,Sports,41.36,28.96,3.77,8.45,76,8.0,E
1,Super Mario Bros.,NES,1985,Platform,29.08,3.58,6.81,0.77,<NA>,<NA>,<NA>
2,Mario Kart Wii,Wii,2008,Racing,15.68,12.76,3.79,3.29,82,8.3,E
3,Wii Sports Resort,Wii,2009,Sports,15.61,10.93,3.28,2.95,80,8.0,E
4,Pokemon Red/Pokemon Blue,GB,1996,Role-Playing,11.27,8.89,10.22,1.0,<NA>,<NA>,<NA>
...,...,...,...,...,...,...,...,...,...,...,...
16951,Samurai Warriors: Sanada Maru,PS3,2016,Action,0.0,0.0,0.01,0.0,<NA>,<NA>,<NA>
16952,LMA Manager 2007,X360,2006,Sports,0.0,0.01,0.0,0.0,<NA>,<NA>,<NA>
16953,Haitaka no Psychedelica,PSV,2016,Adventure,0.0,0.0,0.01,0.0,<NA>,<NA>,<NA>
16954,Spirits & Spells,GBA,2003,Platform,0.01,0.0,0.0,0.0,<NA>,<NA>,<NA>


---
<a class='anchor' id='2-bullet'></a>
## Проверка ошибок в данных и их предобработка

<a class='anchor' id='2-1-bullet'></a>
### Названия, или метки, столбцов датафрейма

In [6]:
print(f'Колонки: {df.columns.tolist()}')

Колонки: ['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales', 'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating']


Уже преобразованы в Чистовой загрузке CSV.

<a class='anchor' id='2-2-bullet'></a>
### Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

In [7]:
# Данная информация устарела: сейчас в pandas есть Extended types для скалярных типов данных и они решают данную проблему пропусков на ура. 
# Преобразование выполнено в Чистовой загрузке CSV.

- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

In [8]:
# Выполнено в Чистовой загрузке CSV.

<a class='anchor' id='2-3-bullet'></a>
### Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [9]:
show(1, 'missing', lambda: display(get_missings(df)))


 1. missing -------------------- 



,N/A,Fraction %
user_score,9268,54.66
critic_score,8714,51.39
rating,6871,40.52
year_of_release,275,1.62
eu_sales,6,0.04
jp_sales,4,0.02
genre,2,0.01
name,2,0.01


- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.


Во всех трёх тейтингах и скорингах 40-55% пропусков. Люди неохотно отправляют оценки. Как решил выше, в выводах по датасету, будем удалять мусор: там, где нет name или genre. В остальных случаях: year_of_release, rating, score - обходить эти строки мимо при анализе.

In [10]:
display(df[df['genre'].isna() | df['name'].isna()]) # это битые безполезные строки, удалим их

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
661,<NA>,GEN,1993,<NA>,1.78,0.53,0.0,0.08,<NA>,<NA>,<NA>
14439,<NA>,GEN,1993,<NA>,0.0,0.0,0.03,0.0,<NA>,<NA>,<NA>


In [11]:
df = df.dropna(subset=['name', 'genre']).reset_index(drop=True) # удаляем как мусор

Посмотрим на строки с пропущенными sales и year

In [12]:
empty_sales_rows = df[df['jp_sales'].isna() | df['eu_sales'].isna()]
display(empty_sales_rows)

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
446,Rhythm Heaven,DS,2008,Misc,0.55,<NA>,1.93,0.13,83,9.0,E
467,Saints Row 2,X360,2008,Action,1.94,0.79,<NA>,0.28,81,8.1,M
801,Dead Rising,X360,2006,Action,1.16,<NA>,0.08,0.2,85,7.6,M
818,UFC 2009 Undisputed,X360,2009,Fighting,1.48,0.39,<NA>,0.19,83,7.9,T
1130,Prince of Persia: Warrior Within,PS2,2004,Action,0.54,<NA>,0.0,0.22,83,8.5,M
1131,Far Cry 4,XOne,2014,Shooter,0.8,<NA>,0.01,0.14,82,7.5,M
1378,Hello Kitty Party,DS,2007,Misc,0.78,0.51,<NA>,0.12,<NA>,<NA>,E
1393,Sonic Advance 3,GBA,2004,Platform,0.74,<NA>,0.08,0.06,79,8.4,E
1611,Ratatouille,DS,2007,Action,0.49,<NA>,0.0,0.14,<NA>,<NA>,<NA>
4731,Castlevania: The Dracula X Chronicles,PSP,2007,Platform,0.22,0.09,<NA>,0.07,80,7.8,T


Строки годные, удалять не будем.

In [13]:
display(df[df['year_of_release'].isna()])

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
183,Madden NFL 2004,PS2,<NA>,Sports,4.26,0.26,0.01,0.71,94,8.5,E
379,FIFA Soccer 2004,PS2,<NA>,Sports,0.59,2.36,0.04,0.51,84,6.4,E
458,LEGO Batman: The Videogame,Wii,<NA>,Action,1.8,0.97,0.0,0.29,74,7.9,E10+
477,wwe Smackdown vs. Raw 2006,PS2,<NA>,Fighting,1.57,1.02,0.0,0.41,<NA>,<NA>,<NA>
611,Space Invaders,2600,<NA>,Shooter,2.36,0.14,0.0,0.03,<NA>,<NA>,<NA>
...,...,...,...,...,...,...,...,...,...,...,...
16607,PDC World Championship Darts 2008,PSP,<NA>,Sports,0.01,0.0,0.0,0.0,43,<NA>,E10+
16639,Freaky Flyers,GC,<NA>,Racing,0.01,0.0,0.0,0.0,69,6.5,T
16683,Inversion,PC,<NA>,Shooter,0.01,0.0,0.0,0.0,59,6.7,M
16693,Hakuouki: Shinsengumi Kitan,PS3,<NA>,Adventure,0.01,0.0,0.0,0.0,<NA>,<NA>,<NA>


Строки годные, но из анализа по годам придётся исключать. Есть бизнес-требование: "Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно". На этом этапе придётся исключать данные строки естествееной фильтрацией, но пока оставим.

- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

In [14]:
# пропуски уже помечены типом pd.NA при Чистовой загрузке CSV, это лучше значений-индикаторов

In [15]:
# заменяем пропущенные Sales на средние значения в группе platform, year_of_release

eu_means = df.groupby(['platform', 'year_of_release'])['eu_sales'].transform('mean')
jp_means = df.groupby(['platform', 'year_of_release'])['jp_sales'].transform('mean')
df['eu_sales'] = df['eu_sales'].fillna(eu_means)
df['jp_sales'] = df['jp_sales'].fillna(jp_means)

# это для пустых групп, чтобы не появилось N/A
df[['eu_sales', 'jp_sales']] = df[['eu_sales', 'jp_sales']].fillna(0)

display(df.loc[empty_sales_rows.index]) # теперь все Sales pfgjkytys

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
446,Rhythm Heaven,DS,2008,Misc,0.55,0.063883,1.93,0.13,83,9.0,E
467,Saints Row 2,X360,2008,Action,1.94,0.79,0.012721,0.28,81,8.1,M
801,Dead Rising,X360,2006,Action,1.16,0.109785,0.08,0.2,85,7.6,M
818,UFC 2009 Undisputed,X360,2009,Fighting,1.48,0.39,0.011322,0.19,83,7.9,T
1130,Prince of Persia: Warrior Within,PS2,2004,Action,0.54,0.213784,0.0,0.22,83,8.5,M
1131,Far Cry 4,XOne,2014,Shooter,0.8,0.291,0.01,0.14,82,7.5,M
1378,Hello Kitty Party,DS,2007,Misc,0.78,0.51,0.07126,0.12,<NA>,<NA>,E
1393,Sonic Advance 3,GBA,2004,Platform,0.74,0.100449,0.08,0.06,79,8.4,E
1611,Ratatouille,DS,2007,Action,0.49,0.087139,0.0,0.14,<NA>,<NA>,<NA>
4731,Castlevania: The Dracula X Chronicles,PSP,2007,Platform,0.22,0.09,0.04306,0.07,80,7.8,T


<a class='anchor' id='2-4-bullet'></a>
### Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

Приведём названия и жанры к lowercase, чтобы была правильная фильтрация и группировка по ним. При этом пропадут неявные дубликаты. Например: ACTION и Action.

In [16]:
df['name'] = df['name'].str.lower()
df['genre'] = df['genre'].str.lower()
# get_reference_info(df, 'name', False)
get_reference_info(df, 'genre', False)

genre:  ['action', 'adventure', 'fighting', 'misc', 'platform', 'puzzle', 'racing', 'role-playing', 'shooter', 'simulation', 'sports', 'strategy'] 12

genre
action          3418
sports          2375
misc            1772
role-playing    1516
shooter         1346
adventure       1323
racing          1273
platform         904
simulation       884
fighting         862
strategy         691
puzzle           590
Name: count, dtype: Int64



- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

In [17]:
implicit_dups = df[df.duplicated(subset=['name', 'platform', 'year_of_release'], keep=False)]
display(implicit_dups.sort_values(by='name'))

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
15300,11eyes: crossover,X360,2009,adventure,0.0,0.0,0.02,0.0,<NA>,<NA>,<NA>
15299,11eyes: crossover,X360,2009,adventure,0.0,0.0,0.02,0.0,<NA>,<NA>,<NA>
4860,18 wheeler: american pro trucker,PS2,2001,racing,0.2,0.15,0.0,0.05,61,5.7,E
4859,18 wheeler: american pro trucker,PS2,2001,racing,0.2,0.15,0.0,0.05,61,5.7,E
13098,4 elements,PC,2009,puzzle,0.0,0.04,0.0,0.01,<NA>,7.4,E
...,...,...,...,...,...,...,...,...,...,...,...
2908,yu-gi-oh! the falsebound kingdom,GC,2002,strategy,0.49,0.13,0.07,0.02,<NA>,<NA>,<NA>
6695,zoo resort 3d,3DS,2011,simulation,0.11,0.09,0.03,0.02,<NA>,<NA>,E
6694,zoo resort 3d,3DS,2011,simulation,0.11,0.09,0.03,0.02,<NA>,<NA>,E
8156,zumba fitness rush,X360,2012,sports,0.0,0.16,0.0,0.02,73,6.2,E10+


- Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

Есть 484 строки, содержащие дубликаты: записи об одной игре одного года выпуска на одной платформе.

In [18]:
df = df.sort_values(by=['name', 'critic_score', 'user_score'], ascending=[True, False, False])
# делаем так, чтобы самая полная строка с данными осталась после удаления дубликатов
df = df.drop_duplicates(subset=['name', 'platform', 'year_of_release'], keep='first').reset_index(drop=True)

- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

In [19]:
rows_before = pd.read_csv("./datasets/new_games.csv").shape[0]
rows_after = df.shape[0]
print(f'Строки до и после: {rows_before} --> {rows_after}: {rows_after - rows_before}')

Строки до и после: 16956 --> 16712: -244


- После проведения предобработки данных напишите общий промежуточный вывод.

Пришлось удалить 244 строки среди которых дубликаты и мусорные строки с безполезными данными (без названия игры, например), так же, нормализовать часть полей и схлопнуть неявные дубликаты в жанрах. Пропущенные данные по Sales удалось выправить, но их немного (всего 10).

---
<a class='anchor' id='3-bullet'></a>
## Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [20]:
df_actual = df.query('year_of_release >= 2000 and year_of_release <= 2013')
print(df_actual.shape[0], ' строк')

12780  строк


---
<a class='anchor' id='4-bullet'></a>
## Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [21]:
df_actual['user_score_cat'] = pd.cut(df_actual['user_score'], bins=[0, 3, 8, 11], labels=["Низкая", "Средняя", "Высокая"], right=False)
df_actual['user_score_cat'] = df_actual['user_score_cat'].astype('string') # для консистентности

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [22]:
df_actual['critic_score_cat'] = pd.cut(df_actual['critic_score'], bins=[0, 30, 80, 101], labels=["Низкая", "Средняя", "Высокая"], right=False)
df_actual['critic_score_cat'] = df_actual['critic_score_cat'].astype('string') # для консистентности

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [23]:
print(df_actual.groupby('user_score_cat', dropna=False)['name'].count(), '\n')
print(df_actual.groupby('critic_score_cat', dropna=False)['name'].count())

user_score_cat
Высокая    2286
Низкая      116
Средняя    4080
<NA>       6298
Name: name, dtype: int64 

critic_score_cat
Высокая    1691
Низкая       55
Средняя    5422
<NA>       5612
Name: name, dtype: int64


- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [24]:
display(df_actual['platform'].value_counts().head(7))

platform
PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1086
GBA      811
Name: count, dtype: Int64

---
<a class='anchor' id='4-bullet'></a>
## Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

Получен очищенный от дубликатов и мусора, нормализованный и с заполненными пропусками Sales срез датасета с 2000-го по 2013-ый годы. Всего в датасете за этот период содержится информация о 12 780 релизах игр на разных платформах. Выполнена кластеризация игр по оценкам пользователей и критиков в новых колонках user_score_cat и crit_score_cat. Так же, установлено, что за данный период времени больше всего игр выпустила платформа PS2: 2127 штук.

Среди рейтингов обнаружился неизвестный рейтинг К-А, который встречается в 3-х записях.